pyrdp/player/gdi/draw.py

```diff
@@ -0,0 +1,702 @@
+#
+# This file is part of the PyRDP project.
+# Copyright (C) 2019 GoSecure Inc.
+# Licensed under the GPLv3 or later.
+#
+
+import logging
+import typing
+from pyrdp.logging import LOGGER_NAMES
+
+from pyrdp.parser.rdp.orders import GdiFrontend
+from pyrdp.parser.rdp.orders.common import Bounds
+from pyrdp.parser.rdp.orders.frontend import BrushStyle, HatchStyle
+from pyrdp.parser.rdp.orders.alternate import CreateOffscreenBitmap, SwitchSurface, CreateNineGridBitmap, \
+    StreamBitmapFirst, StreamBitmapNext, GdiPlusFirst, GdiPlusNext, GdiPlusEnd, GdiPlusCacheFirst, \
+    GdiPlusCacheNext, GdiPlusCacheEnd, FrameMarker
+from pyrdp.parser.rdp.orders.secondary import CacheBitmapV1, CacheBitmapV2, CacheBitmapV3, CacheColorTable, \
+    CacheGlyph, CacheBrush
+from pyrdp.parser.rdp.orders.primary import DstBlt, PatBlt, ScrBlt, DrawNineGrid, MultiDrawNineGrid, \
+    LineTo, OpaqueRect, SaveBitmap, MemBlt, Mem3Blt, MultiDstBlt, MultiPatBlt, MultiScrBlt, MultiOpaqueRect, \
+    FastIndex, PolygonSc, PolygonCb, PolyLine, FastGlyph, EllipseSc, EllipseCb, GlyphIndex, Brush, \
+    BACKMODE_TRANSPARENT
+
+from pyrdp.ui import QRemoteDesktop, RDPBitmapToQtImage
+
+from .cache import BitmapCache, BrushCache, PaletteCache, GlyphCache, GlyphEntry
+from .raster import set_rop3, set_rop2
+
+from PySide2.QtCore import Qt, QPoint
+from PySide2.QtGui import QImage, QPainter, QColor, QPixmap, QBrush, QPen, QPolygon
+
+LOG = logging.getLogger(LOGGER_NAMES.PLAYER + '.gdi')
+
+SCREEN_BITMAP_SURFACE = 0xFFFF
+BITMAPCACHE_SCREEN_ID = 0xFF
+SUBSTITUTE_SURFACE = -1
+
+# Opaque Rectangle Encoding Flags.
+OPRECT_BOTTOM_ABSENT = 0x01
+OPRECT_RIGHT_ABSENT = 0x02
+OPRECT_TOP_ABSENT = 0x04
+OPRECT_LEFT_ABSENT = 0x08
+
+GLYPH_SPECIAL_PROCESSING = -32768
+GLYPH_FRAGMENT_USE = 0xFE
+GLYPH_FRAGMENT_ADD = 0xFF
+
+# Defined in 2.2.2.2.1.1.2.13 (GlyphIndex)
+SO_VERTICAL = 0x01
+SO_HORIZONTAL = 0x02
+SO_CHAR_INC_EQUAL_BM_BASE = 0x20
+
+
+def rgb_to_qcolor(color: int):
+    """Convert an RDP color (0xRRGGBB) to a QColor."""
+    bpp = 16
+    # TODO: check BPP from capabilities
+    if bpp in [24, 32]:
+        return QColor(color)
+    if bpp == 16:
+        t = color & 0x1F
+        t = (t << 3) + t // 4
+        b = min(t, 255)
+
+        t = (color >> 5) & 0x3F
+        t = (t << 2) + t // 4 // 2
+        g = min(t, 255)
+
+        t = (color >> 11) & 0x1F
+        t = (t << 3) + t // 4
+        r = min(t, 255)
+
+    elif bpp == 15:
+        pass
```

Comment:
maybe raise instead to avoid potential future bugs?

Reply:
`QColor.fromRgb()` will raise on undefined symbols already